src/lib/components/upload/UploadSection.tsx

```diff
@@ -37,120 +40,183 @@ export const UploadSection = ({
   const { address } = useWallet();
   const { broadcast } = useTxBroadcast();
   const { updateCodeInfo } = useCodeStore();
+  const postUploadTx = useUploadContractTx(isMigrate);
+  const { validateUserAddress, validateContractAddress } = useValidateAddress();
+
+  const [estimatedFee, setEstimatedFee] = useState<StdFee>();
+  const [simulateError, setSimulateError] = useState<string>();
+
+  const simulateSuccessText = "Valid Wasm file and instantiate permission";
 
   const {
     control,
-    watch,
     setValue,
+    watch,
     formState: { errors },
+    trigger,
   } = useForm<UploadSectionState>({
     defaultValues: {
       wasmFile: undefined,
-      codeDesc: "",
-      estimatedFee: undefined,
-      simulateStatus: "pending",
-      simulateError: "",
+      codeName: "",
+      permission: AccessType.ACCESS_TYPE_EVERYBODY,
+      addresses: [{ address: "" as Addr }],
     },
     mode: "all",
   });
-  const { wasmFile, codeDesc, estimatedFee, simulateStatus, simulateError } =
-    watch();
 
-  const postUploadTx = useUploadContractTx(isMigrate);
+  const { wasmFile, codeName, addresses, permission } = watch();
+
+  const setDefaultBehavior = () => {
+    setSimulateError(undefined);
+    setEstimatedFee(undefined);
+  };
+
+  const isAddressesError = addresses.some((addr) =>
+    Boolean(
+      validateUserAddress(addr.address) && validateContractAddress(addr.address)
+    )
+  );
+
+  // Should not simulate when permission is any of addresses and address input is not filled, invalid, or empty
+  const isAnyAddrWithoutInput =
+    permission === AccessType.ACCESS_TYPE_ANY_OF_ADDRESSES &&
+    (!addresses.some((addr) => addr.address !== "") || isAddressesError);
```

Comment:
We should merge `isAddressesError` and `isAnyAddrWithoutInput`, and change `addr.address !== ""` to `addr.address.trim().length !== 0`